core/src/main/java/com/linecorp/armeria/client/DefaultDnsQueryLifecycleObserver.java

```diff
@@ -0,0 +1,165 @@
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.client;
+
+import java.net.InetSocketAddress;
+import java.util.Arrays;
+import java.util.List;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+import com.google.common.base.Ascii;
+
+import com.linecorp.armeria.common.metric.MeterIdPrefix;
+
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.Tag;
+import io.netty.channel.ChannelFuture;
+import io.netty.handler.codec.dns.DnsQuestion;
+import io.netty.handler.codec.dns.DnsResponseCode;
+import io.netty.resolver.dns.DnsNameResolverTimeoutException;
+import io.netty.resolver.dns.DnsQueryLifecycleObserver;
+
+/**
+ * A {@link DnsQueryLifecycleObserver} that helps capture custom dns metrics.
+ */
+final class DefaultDnsQueryLifecycleObserver implements DnsQueryLifecycleObserver {
+    private static final String NAME_TAG = "name";
+    private static final String RESULT_TAG = "result";
+    private static final String SERVERS_TAG = "servers";
+    private static final String SERVER_TAG = "server";
+    private static final String CODE_TAG = "code";
+    private static final String CAUSE_TAG = "cause";
+    private static final String CNAME_TAG = "cname";
+
+    private static final Pattern NXDOMAIN_EXCEPTION = Pattern.compile("\\bNXDOMAIN\\b");
+    private static final Pattern CNAME_EXCEPTION = Pattern.compile("\\bCNAME\\b/");
+    private static final Pattern NO_MATCHING_EXCEPTION = Pattern.compile("\\bmatching\\b");
+    private static final Pattern UNRECOGNIZED_TYPE_EXCEPTION = Pattern.compile("\\bunrecognized\b/");
+    private static final Pattern NS_EXHAUSTED_EXCEPTION = Pattern.compile("\\bname\\sservers\\b");
+
+    private final MeterRegistry meterRegistry;
+    private final MeterIdPrefix meterIdPrefix;
+    private final DnsQuestion question;
+
+    private enum DnsExceptionTypes {
+
+        NX_DOMAIN,
+        CNAME_NOT_FOUND,
+        NO_MATCHING_RECORD,
+        UNRECOGNIZED_TYPE,
+        NAME_SERVERS_EXHAUSTED,
+        OTHERS,
+        SERVER_TIMEOUT,
+        RESOLVER_TIMEOUT;
+
+        final String lowerCasedName;
+
+        DnsExceptionTypes() {
+            lowerCasedName = Ascii.toLowerCase(name());
+        }
+    }
+
+    /**
+     * Accepts meterRegistry.
+     * @param meterRegistry {@link MeterRegistry} MeterRegistry to capture metrics.
+     * @param question {@link DnsQuestion} DnsQuestion.
+     */
+    DefaultDnsQueryLifecycleObserver(MeterRegistry meterRegistry,
+                                     DnsQuestion question, MeterIdPrefix prefix) {
+        this.meterRegistry = meterRegistry;
+        meterIdPrefix = prefix;
+        this.question = question;
+    }
+
+    @Override
+    public void queryWritten(InetSocketAddress dnsServerAddress, ChannelFuture future) {
+        meterRegistry.counter(meterIdPrefix.name().concat(".written"),
```

Comment:
Let's just use `+` as IDEA suggested which is more readable:
`meterIdPrefix.name() + ".written"`

Reply:
done